From 665f1be75c55bde0e4bd8f9e654446f5b5288cbf Mon Sep 17 00:00:00 2001
From: hakuna77 <194330619+hakuna77@users.noreply.github.com>
Date: Wed, 24 Dec 2025 14:45:03 +0700
Subject: [PATCH] KernelSU, selinux/avc: add slow_avc_audit hook

---
 security/selinux/avc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/security/selinux/avc.c b/security/selinux/avc.c
index 59fa20b53..806dc8534 100644
--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -776,6 +776,10 @@ static void avc_audit_post_callback(struct audit_buffer *ab, void *a)
 	}
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_slow_avc_audit_new(u32 tsid, u16 *tclass);
+#endif
+
 /* This is the slow part of avc audit with big stack footprint */
 noinline int slow_avc_audit(struct selinux_state *state,
 			    u32 ssid, u32 tsid, u16 tclass,
@@ -786,6 +790,11 @@ noinline int slow_avc_audit(struct selinux_state *state,
 	struct common_audit_data stack_data;
 	struct selinux_audit_data sad;
 
+#ifdef CONFIG_KSU
+	ksu_handle_slow_avc_audit_new(tsid, &tclass);
+	if (!tclass)
+		return 0;	
+#endif
 	if (!a) {
 		a = &stack_data;
 		a->type = LSM_AUDIT_DATA_NONE;
-- 
2.51.2

