From 9ddfd74201d9658086fee827eab3bfe1d7871369 Mon Sep 17 00:00:00 2001
From: hakuna77 <194330619+hakuna77@users.noreply.github.com>
Date: Mon, 5 Jan 2026 23:04:44 +0700
Subject: [PATCH] avc happy

---
 kernel/extras.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/kernel/extras.c b/kernel/extras.c
index b4fe4cd8..30b366c4 100644
--- a/kernel/extras.c
+++ b/kernel/extras.c
@@ -157,7 +157,21 @@ static void destroy_kprobe(struct kprobe **kp_ptr)
 	kfree(kp);
 	*kp_ptr = NULL;
 }
-#endif // KSU_KPROBES_HOOK
+#else
+int ksu_handle_slow_avc_audit_new(u32 tsid, u16 *tclass)
+{
+	if (atomic_read(&disable_spoof))
+		return 0;
+
+	if (tsid != su_sid)
+		return 0;
+
+	pr_info("avc_spoof/slow_avc_audit: prevent log for sid: %u\n", su_sid);
+	*tclass = 0;
+
+	return 0;
+}
+#endif // CONFIG_KPROBES
 
 void ksu_avc_spoof_disable(void)
 {
-- 
2.51.2

