From 220351f7c285a399f2e014098b979f529dc25bcf Mon Sep 17 00:00:00 2001
From: hakuna77 <194330619+hakuna77@users.noreply.github.com>
Date: Thu, 4 Dec 2025 23:31:23 +0000
Subject: [PATCH] test

---
 kernel/setuid_hook.c | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/kernel/setuid_hook.c b/kernel/setuid_hook.c
index 2e08a020..09ccd4b7 100644
--- a/kernel/setuid_hook.c
+++ b/kernel/setuid_hook.c
@@ -201,21 +201,18 @@ do_umount:
 	ksu_handle_umount(old_uid, new_uid);
 
 #ifdef CONFIG_KSU_SUSFS
-	get_task_struct(current);
 
 #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 	// We can reorder the mnt_id now after all sus mounts are umounted
 	susfs_reorder_mnt_id();
 #endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 
-	susfs_set_current_proc_umounted();
-
-	put_task_struct(current);
-
 #ifdef CONFIG_KSU_SUSFS_SUS_PATH
 	susfs_run_sus_path_loop(new_uid);
 #endif // #ifdef CONFIG_KSU_SUSFS_SUS_PATH
-#endif
+
+	susfs_set_current_proc_umounted();
+#endif // #ifdef CONFIG_KSU_SUSFS
 	return 0;
 }
 
-- 
2.34.1

