From caa3e36fb8a958ec29a2dea5c0c536db2df3901b Mon Sep 17 00:00:00 2001
From: hakuna77 <194330619+hakuna77@users.noreply.github.com>
Date: Sun, 7 Dec 2025 23:29:55 +0000
Subject: [PATCH] kernel: Use GFP_ATOMIC for atomic context

---
 kernel/selinux/sepolicy.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/selinux/sepolicy.c b/kernel/selinux/sepolicy.c
index d58fdffd..262d8b4d 100644
--- a/kernel/selinux/sepolicy.c
+++ b/kernel/selinux/sepolicy.c
@@ -356,7 +356,7 @@ static void add_xperm_rule_raw(struct policydb *db, struct type_datum *src,
 		if (datum->u.xperms == NULL) {
 			datum->u.xperms =
 				(struct avtab_extended_perms *)(kzalloc(
-					sizeof(xperms), GFP_KERNEL));
+					sizeof(xperms), GFP_ATOMIC));
 			if (!datum->u.xperms) {
 				pr_err("alloc xperms failed\n");
 				return;
-- 
2.34.1

